### PR TITLE
piece-wise linear annealing hooks (loss weight, model attribute, lear…

### PR DESCRIPTION
**Pull request by JanekEbb**
…ning rate)

### CODE REVIEW

File: padertorch/train/hooks.py [hunk 1]
```diff
@@ -812,105 +816,115 @@ class StopTraining(Exception):
...
-    Anneals a loss weight within the los_weights dict of the trainer.
-    """
-    def __init__(self, name, factor, trigger, max_value=None, min_value=None):
+class AnnealingHook(TriggeredHook):
+    def __init__(self, trigger, breakpoints, name):
         """
+        Base class for piece-wise linear annealing. The piece-wise linear
+        function is parameterized by its breakpoints. Can also be used for
+        arbitrary annealing functions by stating breakpoints with an interval
+        similar to the trigger interval.
```
Comment (boeddeker) [inline padertorch/train/hooks.py:825]: Can you add a comment, what happens before the first breakpoint and what after the last breakpoint?

File: padertorch/train/hooks.py [hunk 2]
```diff
@@ -812,105 +816,115 @@ class StopTraining(Exception):
...
         self.name = name
-        self.factor = factor
-        self.max_value = max_value
-        self.min_value = min_value
+        self.last_break = None
 
     @property
     def uid(self):
-        return super().uid + f"({self.name})"
+        return super().uid + f"({'.'.join(self.name)})"
```
Comment (boeddeker) [inline padertorch/train/hooks.py:840]: Why do you use join here? Name should be already a string.

File: padertorch/train/hooks.py [hunk 3]
```diff
@@ -812,105 +816,115 @@ class StopTraining(Exception):
...
+        if self.trigger(iteration=trainer.iteration, epoch=trainer.epoch):
+            if self.last_break is None:
+                self.last_break = (0, self.get_value(trainer))
+            while len(self.breakpoints) > 0 and self.breakpoints[0][0] <= trainer.iteration:
+                self.last_break = self.breakpoints.pop(0)
+            if len(self.breakpoints) > 0:
+                slope = (self.breakpoints[0][1]-self.last_break[1])/(self.breakpoints[0][0]-self.last_break[0])
+                value = self.last_break[1] + slope * (trainer.iteration - self.last_break[0])
+            else:
+                value = self.last_break[1]
```
Comment (boeddeker) [inline padertorch/train/hooks.py:858]: This code is relative complicated.
Is my understanding correct: Take the original value as start values and then interpolate to the next break point?
i.e. the breakpoints are absolute and not relative the original value.

File: padertorch/train/hooks.py [hunk 4]
```diff
@@ -812,105 +816,115 @@ class StopTraining(Exception):
...
+            trigger:
+            breakpoints:
+            name: states the key of the target optimizer when optimizer is a dict
+        """
+        super().__init__(trigger, breakpoints, name)
+
+    @property
+    def uid(self):
+        if self.name is None:
+            return super(AnnealingHook, self).uid
```
Comment (boeddeker) [inline padertorch/train/hooks.py:913]: Can you add a comment, that you didn't write python 2 code here?

File: padertorch/train/hooks.py (hunk 1 above)
Comment (JanekEbb) [reply padertorch/train/hooks.py:825]: done

File: padertorch/train/hooks.py (hunk 2 above)
Comment (JanekEbb) [reply padertorch/train/hooks.py]: fixed

File: padertorch/train/hooks.py (hunk 3 above)
Comment (JanekEbb) [reply padertorch/train/hooks.py:867]: right, I added comments to improve clarity of the math

File: padertorch/train/hooks.py (hunk 4 above)
Comment (JanekEbb) [reply padertorch/train/hooks.py]: done

**Comment by boeddeker** [comment]
Can you explain, why you used absolute values for the break points?
I expected relative values.

**Comment by JanekEbb** [comment]
> Can you explain, why you used absolute values for the break points?
> I expected relative values.

Hm, not really. Maybe because its more verbose which values are used. Can you explain why you expected relative values?

**Comment by boeddeker** [comment]
> Hm, not really. Maybe because its more verbose which values are used. Can you explain why you expected relative values?

When you read about learning rate decay, the people often write something like: after X epochs/iterations I reduce the learning rate by a factor of Y.

**Comment by JanekEbb** [comment]
> > Hm, not really. Maybe because its more verbose which values are used. Can you explain why you expected relative values?
> 
> When you read about learning rate decay, the people often write something like: after X epochs/iterations I reduce the learning rate by a factor of Y.

so you expected each breakpoint to be relative to the previous breakpoint? Anyway one reason to not use relative breakpoints is to allow ramp ups from 0 to some value. I added a comment.

**Comment by boeddeker** [comment]
> so you expected each breakpoint to be relative to the previous breakpoint? Anyway one reason to not use relative breakpoints is to allow ramp ups from 0 to some value. I added a comment.

I expected more, relative to the initial value. When I change the learning rate, I must now also change the breakpoints. With relative to the initial value are also ramps from zero possible.

**Comment by JanekEbb** [comment]
> > so you expected each breakpoint to be relative to the previous breakpoint? Anyway one reason to not use relative breakpoints is to allow ramp ups from 0 to some value. I added a comment.
> 
> I expected more, relative to the initial value. When I change the learning rate, I must now also change the breakpoints. With relative to the initial value are also ramps from zero possible.

When the inital value is 0 !?

**Comment by boeddeker** [comment]
> When the inital value is 0 !?

When the default value is 0, that value is deactivated. But why not interpret this value as "scale"?
With
Breakpoint: (0, 0), (10, 1), (20, 0)
and an initial value ~0 you also get a ramp starting from zero.

**Comment by boeddeker** [comment]
What is the opinion of the others (@jensheit @thequilo @tglarner @TCord)?
Is a relative or absolute value more practical/intuitive?

**Comment by thequilo** [comment]
For me, absolute values are more intuitive. I don't know how this is used in the literature so this might not be true for everyone. I also agree that relative values can be more practical in some cases. How about supporting both?

**Comment by JanekEbb** [comment]
> > When the inital value is 0 !?
> 
> When the default value is 0, that value is deactivated. But why not interpret this value as "scale"?
> With
> Breakpoint: (0, 0), (10, 1), (20, 0)
> and an initial value ~0 you also get a ramp starting from zero.

Now I see your point. Agree.

**Comment by JanekEbb** [comment]
> For me, absolute values are more intuitive. I don't know how this is used in the literature so this might not be true for everyone. I also agree that relative values can be more practical in some cases. How about supporting both?

Sorry saw your comment only after the latest push. Do you think it is enough that the annealing function corresponds to absolute values if the initial value is set to 1 ?

**Comment by thequilo** [comment]
Yes, I guess that's fine

File: padertorch/train/hooks.py [hunk 5]
```diff
@@ -812,105 +816,136 @@ class StopTraining(Exception):
...
+
+
+class LRAnnealingHook(AnnealingHook):
+    """
+    Anneals an optimizer learning rate.
+    """
+    def __init__(self, trigger, breakpoints, name=None):
+        """
 
+        Args:
```
Comment (boeddeker) [inline padertorch/train/hooks.py:924]: Can you either copy the docstring from AnnealingHook or mention that the signature description is in AnnealingHook?

**Comment by boeddeker** [comment]
One last question: The breakpoint unit is iteration. Should the unit epoch also be supported?

**Comment by thequilo** [comment]
I'd say yes, because with iterations we have to change the breakpoints when the batch-size changes. Also, I think that many papers use epoch to describe such an annealing.

**Comment by JanekEbb** [comment]
Do you suggest to adopt the unit from trigger or to add a separate argument?

**Comment by boeddeker** [comment]
> Do you suggest to adopt the unit from trigger or to add a separate argument?

Yes and yes.

I vote for a separate `unit` argument. i.e.
```python
breakpoints=[(1, 3), (2, 3)], unit='epoch'
```
and not
```python
breakpoints=[((1, 'epoch'), 3), ((2, 'epoch'), 3)]
```
(Do you mean something like that with `adopt the unit from trigger`?)

I think we should not support different units for the breakpoints.
That would make the implementation complicated and I do not see a usecase.

File: padertorch/train/hooks.py [hunk 6]
```diff
@@ -812,105 +816,144 @@ class StopTraining(Exception):
...
+            assert isinstance(optimizer, dict)
+            optimizer = optimizer[self.name]
+        return optimizer.optimizer
+
+    def get_value(self, trainer):
+        return self.get_optimizer(trainer).param_groups[0]['lr']
+
+    def set_value(self, trainer, value):
+        optimizer = self.get_optimizer(trainer)
+        for param_group in optimizer.param_groups:
```
Comment (thequilo) [inline padertorch/train/hooks.py:958]: Does this work if the optimizer is a `dict` but `self.name=None`?

Comment (thequilo) [reply padertorch/train/hooks.py:958]: If not, you should assert this in `get_optimizer`

File: padertorch/train/hooks.py [hunk 7]
```diff
@@ -946,8 +946,13 @@ def uid(self):
...
         optimizer = trainer.optimizer
         if self.name is not None:
-            assert isinstance(optimizer, dict)
+            assert isinstance(optimizer, dict), type(optimizer)
             optimizer = optimizer[self.name]
+        else:
+            assert (
+                hasattr(optimizer, 'optimizer')
+                and hasattr(optimizer.optimizer, 'param_groups')
+            ), type(optimizer)
```
Comment (thequilo) [inline padertorch/train/hooks.py:955]: This assert can as well be performed when `self.name is not None:`, the same assertion should hold for the contents of the optimizer dict. Can you remove the `else:`?
